### PR TITLE
Updates/merge multiple policy

### PR DESCRIPTION
**Pull request by r-scheele**
## Description
This PR merges multiple policies in `mc admin user policy TARGET USERNAME`    command, if more than one exists.

```bash
./mc admin user add myminio minio1 minio123
Added user `minio1` successfully.
```

```bash
$ ./mc admin policy list myminio
consoleAdmin
diagnostics
readonly
readwrite
writeonly
```

```bash
$ ./mc admin user policy myminio minio1
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "admin:*"
      ]
    },
    {
      "Effect": "Allow",
      "Action": [
        "kms:*"
      ]
    },
    {
      "Effect": "Allow",
      "Action": [
        "s3:*"
      ],
      "Resource": [
        "arn:aws:s3:::*"
      ]
    }
  ]
}
```

### CODE REVIEW

File: cmd/admin-user-policy.go [hunk 1]
```diff
@@ -76,12 +78,33 @@ func mainAdminUserPolicy(ctx *cli.Context) error {
...
+
+	var policies []policy.Policy
+	for _, policyName := range policyNames {
+		policyInfo, err := getPolicyInfo(client, policyName)
+		if err != nil {
+			fatalIf(probe.NewError(err).Trace(), "Unable to fetch user policy document for policy "+policyName)
+			return nil
+		}
+
+		if len(policyNames) == 1 {
```
Comment (shtripat) [inline cmd/admin-user-policy.go:91]: This check can be done before for loop and print only policy and return.

File: cmd/admin-user-policy.go [hunk 2]
```diff
@@ -76,12 +78,34 @@ func mainAdminUserPolicy(ctx *cli.Context) error {
...
-		fmt.Println(string(pinfo.Policy))
+	policyNames := strings.Split(user.PolicyName, ",")
+
+	if len(policyNames) == 1 {
+		policyInfo, err := getPolicyInfo(client, policyNames[0])
+		if err != nil {
+			fatalIf(probe.NewError(err).Trace(), "Unable to fetch user policy document")
+		}
+		fmt.Println(string(policyInfo.Policy))
+		return nil
```
Comment (harshavardhana) [inline cmd/admin-user-policy.go:89]:
```suggestion
	if len(policyNames) == 1 {
		policyInfo, err := getPolicyInfo(client, policyNames[0])
		if err != nil {
			fatalIf(probe.NewError(err).Trace(), "Unable to fetch user policy document")
		}
		fmt.Println(string(policyInfo.Policy))
		return nil
```

This is not needed not sure why we have this.

File: cmd/admin-user-policy.go [hunk 3]
```diff
@@ -76,12 +78,34 @@ func mainAdminUserPolicy(ctx *cli.Context) error {
...
 	}
 
+	var policies []policy.Policy
+	for _, policyName := range policyNames {
```
Comment (harshavardhana) [inline cmd/admin-user-policy.go:93]:
```suggestion
	for _, policyName := range policyNames {
	         if policyName == "" {
	              continue
	         }
	}
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/admin-user-policy.go (hunk 2 above)
Comment (r-scheele) [reply cmd/admin-user-policy.go:89]: Thanks for the suggestion @harshavardhana `MergePolicies()` works for a policy too

File: cmd/admin-user-policy.go (hunk 3 above)
Comment (r-scheele) [reply cmd/admin-user-policy.go:84]: Hi @harshavardhana i already submitted a fix for that in the next commit

**Review by harshavardhana — APPROVED** [review]

**Review by donatello — CHANGES_REQUESTED** [review]
I think output should indicate that this is a merged policy output of the policies applied to the user to reduce chances of confusion.

File: cmd/admin-user-policy.go [hunk 4]
```diff
@@ -76,12 +78,28 @@ func mainAdminUserPolicy(ctx *cli.Context) error {
...
-		fatalIf(probe.NewError(e).Trace(args...), "Unable to fetch user policy document for policy "+policy)
-		fmt.Println(string(pinfo.Policy))
+	policyNames := strings.Split(user.PolicyName, ",")
+
+	var policies []policy.Policy
+	for _, policyName := range policyNames {
+		if policyName == "" {
+			continue
+		}
+		policyInfo, err := getPolicyInfo(client, policyName)
```
Comment (harshavardhana) [inline cmd/admin-user-policy.go:88]:
```suggestion
		policyInfo, e := getPolicyInfo(client, policyName)
		fatalIf(probe.NewError(e).Trace(), "Unable to fetch user policy document for policy "+policyName)
```

File: cmd/admin-user-policy.go [hunk 5]
```diff
@@ -76,12 +78,28 @@ func mainAdminUserPolicy(ctx *cli.Context) error {
...
+		if err != nil {
+			fatalIf(probe.NewError(err).Trace(), "Unable to fetch user policy document for policy "+policyName)
+		}
+
+		var policyObj policy.Policy
+		if err := json.Unmarshal(policyInfo.Policy, &policyObj); err != nil {
```
Comment (harshavardhana) [inline cmd/admin-user-policy.go:94]:
```suggestion
		if e := json.Unmarshal(policyInfo.Policy, &policyObj); e != nil {
```

File: cmd/admin-user-policy.go [hunk 6]
```diff
@@ -76,12 +78,28 @@ func mainAdminUserPolicy(ctx *cli.Context) error {
...
+			fatalIf(probe.NewError(err).Trace(), "Unable to unmarshal policy")
```
Comment (harshavardhana) [inline cmd/admin-user-policy.go:95]:
```suggestion
			fatalIf(probe.NewError(e).Trace(), "Unable to unmarshal policy")
```

File: cmd/admin-user-policy.go [hunk 7]
```diff
@@ -76,12 +78,28 @@ func mainAdminUserPolicy(ctx *cli.Context) error {
...
+		}
+		policies = append(policies, policyObj)
 	}
 
+	mergedPolicy := policy.MergePolicies(policies...)
+	policyStr, _ := json.Marshal(mergedPolicy)
```
Comment (harshavardhana) [inline cmd/admin-user-policy.go:101]:
```suggestion
	policyStr, _ := json.Marshal(mergedPolicy)
```

don't ignore errors or use `json.NewEncoder(os.Stdout).Encode(mergedPolicy)`

**Review by harshavardhana — CHANGES_REQUESTED** [review]
In mc we treat `go` err as `e` and `probe.Error` as `err`

File: cmd/admin-user-policy.go (hunk 7 above)
Comment (r-scheele) [reply cmd/admin-user-policy.go:101]: This makes sense, thanks @harshavardhana

**Review by harshavardhana — APPROVED** [review]

**Comment by harshavardhana** [comment]
PTAL @donatello

**Comment by harshavardhana** [comment]
PTAL @donatello @shtripat

**Review by shtripat — APPROVED** [review]
lgtm

**Pull request merged**